docs/proposal/20210311-single-controller-multitenancy.md

```diff
@@ -0,0 +1,321 @@
+
+Changed Resources
+
+- `VsphereCluster`
+
+New Resources
+
+- cluster-scoped `VsphereClusterAccount` represents the vCenter account to use for reconciliation. This type should also contain references to namespaces that are allowed to use the account.
+
+Changes to VsphereCluster
+A new field is added to the VsphereClusterSpec to reference the `VsphereClusterAccount`. We intend to use `corev1.TypedLocalObjectReference` in order to ensure the only objects that can be referenced are either in the same namespace or are scoped globally.
+
+```go
+// VsphereClusterAccount is the account to be used for vcenter actions
+type VsphereClusterAccount struct {
+    metav1.TypeMeta `json:`",inline"`
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    Spec VsphereClusterAccountSpec `json:"spec,omitempty"`
+    Status VsphereClusterAccountStatus `json:"status, omitempty"`
+}
+
+type VsphereClusterAccountSpec struct {
+    // CredentialsRef references a Secret with the credentials to use
+    CredentialsRef *corev1.ObjectReference `json:"credentialsRef,omitempty"`
+
+    // AllowedNamespaces is used to identify which namespaces are allowed to use this account.
+    // Namespaces can be selected either by using an array of namespaces or with a label selector.
+    // If this object is nil, no namespaces will be allowed
+    //
+    // +optional
+    AllowedNamespaces *AllowedNamespaces `json:"allowedNamespaces"`
+}
+
+type AllowedNamespaces struct {
+    // If both NamespaceList and Selector are nil/empty, no namespaces will be allowed to use the account (default behaviour)
+    // If both NamespaceList and Selector are defined, the union of the two would be used to determine if a namespace is allowed to use the account
+    //
+    // Allowed namespaces listed out by name
+    // +optional
+    // +nullable
+    NamespaceList []string `json:"list"`
+
+    // Selector is a standard Kubernetes LabelSelector. A label query over a set of resources.
+    // +optional
+    Selector metav1.LabelSelector `json:"selector"`
+}
+
+type VsphereClusterSpec struct {
+    ...
+    // +optional
+    IdentityRef *corev1.TypedLocalObjectReference `json:"IdentityRef,omitempty"`
+}
+```
+
+#### Controller Changes
+
+- If IdentityRef is specified, and it references a `VsphereClusterAccount`, the CRD is fetched and used to create a session
+- If IdentityRef is specified, and it references a `Secret`, the `Secret` is used to create a session
+- The controller will add the VsphereCluster to OwnerReferences on a Secret that contains the account credentials
+- If IdentityRef is not specified, the controller will fallback to using credentials provided in the controller deployment
+- The session will be cached using the existing logic where the key is server + username + datacenter
+- The `IdentityRef` on a `VsphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials
```

Comment:
```suggestion
- The `IdentityRef` on a `VSphereCluster` will be mutable. Setting the `IdentityRef` to nil will cause the controller to fallback to the static credentials
```

Reply:
Consider https://github.com/kubernetes-sigs/cluster-api-provider-aws/blob/main/docs/proposal/20200506-single-controller-multitenancy.md#upgrade-strategy for fallback behaviour.